File: Source/AlphaTab/Rendering/Effects/TempoEffectInfo.cs
using AlphaTab.Model;
using AlphaTab.Rendering.Glyphs;

namespace AlphaTab.Rendering.Effects
{
    internal class TempoEffectInfo : IEffectBarRendererInfo
    {
        public string EffectId => "tempo";
        public bool HideOnMultiTrack => true;
        public bool CanShareBand => false;
        public EffectBarGlyphSizing SizingMode => EffectBarGlyphSizing.SinglePreBeat;

        public bool ShouldCreateGlyph(Settings settings, Beat beat)
        {
            return !settings.Notation.HideBeatInfo && beat.Voice.Bar.Staff.Index == 0 && beat.Voice.Index == 0 && beat.Index == 0 &&
                   (beat.Voice.Bar.MasterBar.TempoAutomation != null || beat.Voice.Bar.Index == 0);
        }

        public EffectGlyph CreateNewGlyph(BarRendererBase renderer, Beat beat)
        {
            int tempo;
            if (beat.Voice.Bar.MasterBar.TempoAutomation != null)
            {
                tempo = (int)beat.Voice.Bar.MasterBar.TempoAutomation.Value;
            }
            else
            {
                tempo = beat.Voice.Bar.Staff.Track.Score.Tempo;
            }

            return new TempoGlyph(0, 0, tempo);
        }

        public bool CanExpand(Beat from, Beat to)
        {
            return true;
        }
    }
}

File: src/rendering/effects/TempoEffectInfo.ts
import type { Beat } from '@src/model/Beat';
import type { BarRendererBase } from '@src/rendering/BarRendererBase';
import { EffectBarGlyphSizing } from '@src/rendering/EffectBarGlyphSizing';
import type { EffectGlyph } from '@src/rendering/glyphs/EffectGlyph';
import { EffectBarRendererInfo } from '@src/rendering/EffectBarRendererInfo';
import type { Settings } from '@src/Settings';
import { NotationElement } from '@src/NotationSettings';
import { BarTempoGlyph } from '@src/rendering/glyphs/BarTempoGlyph';

export class TempoEffectInfo extends EffectBarRendererInfo {
    public get notationElement(): NotationElement {
        return NotationElement.EffectTempo;
    }

    public get hideOnMultiTrack(): boolean {
        return true;
    }

    public get canShareBand(): boolean {
        return false;
    }

    public get sizingMode(): EffectBarGlyphSizing {
        return EffectBarGlyphSizing.SinglePreBeat;
    }

    public shouldCreateGlyph(settings: Settings, beat: Beat): boolean {
        return (
            beat.voice.bar.staff.index === 0 &&
            beat.voice.index === 0 &&
            beat.index === 0 &&
            beat.voice.bar.masterBar.tempoAutomations.length > 0
        );
    }

    public createNewGlyph(renderer: BarRendererBase, beat: Beat): EffectGlyph {
        return new BarTempoGlyph(beat.voice.bar.masterBar.tempoAutomations);
    }

    public canExpand(from: Beat, to: Beat): boolean {
        return true;
    }
}
